### ISSUE TITLE
Redis _ci_redis_serialized

### ISSUE THREAD

**Issue by loga4** [bcit-ci/CodeIgniter#4184]
Hi!
I have some trouble with Redis (use as Cache driver) and "_ci_redis_serialized". It grows very quickly

So, if "_ci_redis_serialized" is so big, then Redis::get is very slow.

**Comment by narfbg** [comment]
Post this on the [forums](http://forum.codeigniter.com/) please, we only accept bug reports here.

**Comment by robotokiller** [comment]
The variable "_ci_redis_serialized" is very bad for performance. I have 1 million keys at Redis it takes very long time to initialize Cache_redis.php.
```php
class CI_Cache_redis extends CI_Driver
{
 public function __construct()
 {
   ...
  // Initialize the index of serialized values.
  $serialized = $this->_redis->sMembers('_ci_redis_serialized');
  empty($serialized) OR $this->_serialized = array_flip($serialized);
 }
}
```

**Comment by narfbg** [comment]
Then store raw values instead of arrays and objects?

It's a trade-off, some functionalities will obviously impact performance ... you can't have everything.

**Comment by robotokiller** [comment]
You are right. Storing primitive data as serialized string takes memory overhead. 
I extend CI_Cache_driver to solve my problem.
File **application/libraries/Cache/drivers/MY_Cache_redis.php**
```php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

class MY_Cache_redis extends CI_Cache_redis
{

    /**
     * Class constructor
     *
     * Setup Redis
     *
     * Loads Redis config file if present. Will halt execution
     * if a Redis connection can't be established.
     *
     * @return	void
     * @see		Redis::connect()
     */
    public function __construct()
    {
        $config = array();
        $CI =& get_instance();

        if ($CI->config->load('redis', TRUE, TRUE))
        {
            $config = $CI->config->item('redis');
        }

        $config = array_merge(self::$_default_config, $config);
        $this->_redis = new Redis();

        try
        {
            if ($config['socket_type'] === 'unix')
            {
                $success = $this->_redis->connect($config['socket']);
            }
            else // tcp socket
            {
                $success = $this->_redis->connect($config['host'], $config['port'], $config['timeout']);
            }

            if ( ! $success)
            {
                log_message('error', 'Cache: Redis connection failed. Check your configuration.');
            }

            if (isset($config['password']) && ! $this->_redis->auth($config['password']))
            {
                log_message('error', 'Cache: Redis authentication failed.');
            }
        }
        catch (RedisException $e)
        {
            log_message('error', 'Cache: Redis connection refused ('.$e->getMessage().')');
        }
        $this->_redis->setOption(Redis::OPT_SERIALIZER, $this->serializer());
    }


    /**
     * Get cache
     *
     * @param	string	Cache ID
     * @return	mixed
     */
    public function get($key)
    {
        return $this->_redis->get($key);
    }

    /**
     * Save cache
     *
     * @param	string	$key	Cache ID
     * @param	mixed	$data	Data to save
     * @param	int	$ttl	Time to live in seconds
     * @param	bool	$raw	Whether to store the raw value (unused)
     * @return	bool	TRUE on success, FALSE on failure
     */
    public function save($key, $data, $ttl = 60, $raw = FALSE)
    {
        return $this->_redis->set($key, $data, $ttl);
    }

    /**
     * Delete from cache
     *
     * @param	string	Cache key
     * @return	bool
     */
    public function delete($key)
    {
        if ($this->_redis->delete($key) !== 1)
        {
            return FALSE;
        }
        return TRUE;
    }

    /**
     * Returns the serializer constant to use.
     *
     * @return int
     */
    protected function serializer(){
        if (defined('Redis::SERIALIZER_IGBINARY')){
            return Redis::SERIALIZER_IGBINARY;
        }
        return Redis::SERIALIZER_PHP;
    }
}
```

**Comment by exussum12** [comment]
as a note this is also fixed with #5739
